backend/danswer/connectors/teams/connector.py

```diff
@@ -0,0 +1,269 @@
+    # Extract the 'createdDateTime' value from the 'properties' dictionary
+    created_datetime_str = obj.properties["createdDateTime"]
+
+    # Convert the string to a datetime object
+    return datetime.strptime(created_datetime_str, "%Y-%m-%dT%H:%M:%S.%f%z")
+
+
+class TeamsConnector(LoadConnector, PollConnector):
+    def __init__(
+        self,
+        batch_size: int = INDEX_BATCH_SIZE,
+        teams: list[str] = [],
+    ) -> None:
+        self.batch_size = batch_size
+        self.graph_client: GraphClient | None = None
+        self.requested_team_list: list[str] = teams
+
+    def load_credentials(self, credentials: dict[str, Any]) -> dict[str, Any] | None:
+        teams_client_id = credentials["teams_client_id"]
+        teams_client_secret = credentials["teams_client_secret"]
+        teams_directory_id = credentials["teams_directory_id"]
+
+        def _acquire_token_func() -> dict[str, Any]:
+            """
+            Acquire token via MSAL
+            """
+            authority_url = f"https://login.microsoftonline.com/{teams_directory_id}"
+            app = msal.ConfidentialClientApplication(
+                authority=authority_url,
+                client_id=teams_client_id,
+                client_credential=teams_client_secret,
+            )
+            token = app.acquire_token_for_client(
+                scopes=["https://graph.microsoft.com/.default"]
+            )
+            return token
+
+        self.graph_client = GraphClient(_acquire_token_func)
+        return None
+
+    def get_post_message_lists_from_channel(
+        self, channel_object: Channel
+    ) -> list[list[ChatMessage]]:
+        base_message_list: list[
+            ChatMessage
+        ] = channel_object.messages.get().execute_query()
+
+        post_message_lists: list[list[ChatMessage]] = []
+        for message in base_message_list:
+            replies = message.replies.get_all().execute_query()
+
+            post_message_list: list[ChatMessage] = [message]
+            post_message_list.extend(replies)
+
+            post_message_lists.append(post_message_list)
+
+        return post_message_lists
+
+    def get_channel_object_list_from_team_list(
+        self,
+        team_object_list: list[Team],
+        start: datetime | None = None,
+        end: datetime | None = None,
+    ) -> list[Channel]:
+        filter_str = ""
+        if start is not None and end is not None:
+            filter_str = f"last_modified_datetime ge {start.isoformat()} and last_modified_datetime le {end.isoformat()}"
+
+        channel_list: list[Channel] = []
+        for team_object in team_object_list:
+            query = team_object.channels.get()
+            if filter_str:
+                query = query.filter(filter_str)
+            channel_objects = query.execute_query()
+            channel_list.extend(channel_objects)
+
+        return channel_list
+
+    def get_all_team_objects(self) -> list[Team]:
```

Comment:
Not that important but across the board, if it's a function that should only be used locally from this file, start it with an underscore _get_all_team_objects() (but also given the advice on naming, this should just be _get_all_teams as the function name)